core/services/wifi/Hotspot.py

```diff
@@ -0,0 +1,339 @@
+
+        binary_path = shutil.which(self.binary_name())
+        if binary_path is None:
+            raise ValueError("Hostapd binary not found on system's PATH.")
+
+        self._binary = pathlib.Path(binary_path)
+        assert self.is_binary_working()
+
+    @staticmethod
+    def binary_name() -> str:
+        return "hostapd"
+
+    def binary(self) -> pathlib.Path:
+        return self._binary
+
+    def set_ssid(self, ssid: str) -> None:
+        logger.debug(f"Changing hotspot ssid to '{ssid}'.")
+        self._ap_ssid = ssid
+
+    def set_passphrase(self, passphrase: str) -> None:
+        logger.debug(f"Changing hotspot passphrase to '{passphrase}'.")
+        self._ap_passphrase = passphrase
+
+    def is_binary_working(self) -> bool:
+        if self.binary() is None:
+            return False
+
+        try:
+            subprocess.check_output([self.binary(), "-h"])
+            return True
+        except subprocess.CalledProcessError as error:
+            if error.returncode == 1:
+                return True
+            logger.error(f"Invalid binary: {error}")
+            return False
+
+    def base_interface_channel_frequency(self) -> int:
+        wireless_interfaces = self.iw.get_interfaces_dict()
+        if not self._base_interface in wireless_interfaces:
+            raise RuntimeError("Could not find base interface.")
+        last_channel = -1
+        time_last_channel_change = time.time()
+        time_start_searching = time.time()
+        while True:
+            current_channel = int(wireless_interfaces[self._base_interface][3])
+            if current_channel != last_channel:
+                time_last_channel_change = time.time()
+                last_channel = current_channel
+            seconds_in_same_channel = time.time() - time_last_channel_change
+            seconds_searching = time.time() - time_start_searching
+            if seconds_in_same_channel > 2:
+                return current_channel
+            if seconds_searching > 15:
+                raise RuntimeError("Could not find base interface channel. Timeout exceeded.")
+
+    def desired_channel_frequency(self) -> int:
+        return self.base_interface_channel_frequency()
+
+    def _create_virtual_interface(self) -> None:
+        logger.debug("Deleting virtual access point interface (if exists).")
+        wireless_interfaces = self.iw.get_interfaces_dict()
+        if self._ap_interface_name in wireless_interfaces:
+            interface_index = int(self.ipr.link_lookup(ifname=self._ap_interface_name)[0])
+            self.iw.del_interface(interface_index)
+        time_start_delete = time.time()
+        while True:
+            if self._ap_interface_name not in self.iw.get_interfaces_dict():
+                time.sleep(0.3)
+                break
+            if time.time() - time_start_delete > 5:
+                raise RuntimeError("Could not delete virtual interface. Timeout exceeded.")
+            time.sleep(0.01)
```

Comment:
Since this code replicates a lot, I would recommend to create a function for it.
```python
def loop_raise_check(wait_condition, raise_message):
    time_start_start = time.time()
    while True:
        if wait_condition():
            break
        if time.time() - time_start_start > 5:
            raise RuntimeError(message)
        time.sleep(0.5)
```